torch_optimizer/adahessian.py

```diff
@@ -0,0 +1,183 @@
+            running averages of gradient and its square (default: (0.9, 0.999))
+        eps (float, optional): term added to the denominator to improve
+            numerical stability (default: 1e-4)
+        weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
+        hessian_power (float, optional): Hessian power (default: 0.5)
+        single_gpu (Bool, optional): Do you use distributed training or not "torch.nn.parallel.DistributedDataParallel" (default: True)
+
+        Example:
+	>>> import torch_optimizer as optim 
+	>>> optimizer = optim.Adahessian(model.parameters(), lr = 1.0)
+	>>> optimizer.zero_grad()
+	>>> loss_fn(model(input), target).backward(create_graph = True) 
+	>>> optimizer.step()
+
+        __ https://arxiv.org/abs/2006.00719
+
+        Note:
+            Reference code: https://github.com/amirgholami/adahessian
+    """
+
+    def __init__(self, params, lr=0.15, betas=(0.9, 0.999), eps=1e-4,
+                 weight_decay=0, hessian_power=0.5, single_gpu=True):
+        if not 0.0 <= lr:
+            raise ValueError("Invalid learning rate: {}".format(lr))
+        if not 0.0 <= eps:
+            raise ValueError("Invalid epsilon value: {}".format(eps))
+        if not 0.0 <= betas[0] < 1.0:
+            raise ValueError(
+                "Invalid beta parameter at index 0: {}".format(
+                    betas[0]))
+        if not 0.0 <= betas[1] < 1.0:
+            raise ValueError(
+                "Invalid beta parameter at index 1: {}".format(
+                    betas[1]))
+        if not 0.0 <= hessian_power <= 1.0:
+            raise ValueError("Invalid Hessian power value: {}".format(hessian_power))
+        defaults = dict(lr=lr, betas=betas, eps=eps,
+                        weight_decay=weight_decay, hessian_power=hessian_power)
+        self.single_gpu = single_gpu 
+        super(Adahessian, self).__init__(params, defaults)
+
+    def get_trace(self, params, grads):
+        """
+        compute the Hessian vector product with a random vector v, at the current gradient point,
+        i.e., compute the gradient of <gradsH,v>.
+        :param gradsH: a list of torch variables
+        :return: a list of torch tensors
+        """
+
+        # Check backward was called with create_graph set to True
+        for i, grad in enumerate(grads):
+            if grad.grad_fn is None:
+                raise RuntimeError('Gradient tensor {:} does not have grad_fn. When calling\n'.format(i) +
+                           '\t\t\t  loss.backward(), make sure the option create_graph is\n' +
+                           '\t\t\t  set to True.')
+
+        v = [2 * torch.randint_like(p, high=2) - 1 for p in params]
+
+        # this is for distributed setting with single node and multi-gpus, 
+        # for multi nodes setting, we have not support it yet.
+        if not self.single_gpu:
+            for v1 in v:
+                dist.all_reduce(v1)
```

Comment:
I think this variable is used before declaration, I guess better to remove multi gpu support, since it is broken anyway.

Reply:
Do not see where `dist` is declared, this is why I think multi gpu possible not working